src/main/java/edu/njit/jerse/automation/ProjectRootFinder.java

```diff
@@ -0,0 +1,90 @@
+package edu.njit.jerse.automation;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.FileVisitResult;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * The {@code ProjectRootFinder} class is responsible for scanning a given directory
+ * and its subdirectories to identify Java project roots. It primarily looks for directories
+ * that conform to standard Java source directory structures, excluding test directories.
+ */
+public class ProjectRootFinder {
+    private static final Logger LOGGER = LogManager.getLogger(ProjectRootFinder.class);
+
+    /**
+     * Finds Java project roots in the given directory and its subdirectories.
+     *
+     * @param directory the root directory from which the search begins
+     * @return a list of {@code File} objects, each representing a discovered Java project root
+     */
+    public static List<File> findJavaRoots(File directory) throws IOException {
+        LOGGER.debug("Searching for Java roots in directory: {}", directory.getAbsolutePath());
+
+        List<File> javaRoots = new ArrayList<>();
+
+        Files.walkFileTree(directory.toPath(), new SimpleFileVisitor<>() {
+            @Override
+            public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) {
+                File dirFile = dir.toFile();
+
+                if (isTestDirectory(dirFile)) {
+                    LOGGER.info("Skipping test directory: {}", dirFile.getAbsolutePath());
+                    // Skip test directories
+                    return FileVisitResult.SKIP_SUBTREE;
+                }
+
+                if (isJavaSourceRoot(dirFile)) {
+                    LOGGER.info("Found Java root: {}", dirFile.getAbsolutePath());
+                    javaRoots.add(dirFile);
+                    // Once a Java source root is found, skip its subdirectories
+                    return FileVisitResult.SKIP_SUBTREE;
+                }
+
+                return FileVisitResult.CONTINUE;
+            }
+
+            @Override
```

Comment:
Should you omit this override?  The default implementation will throw an exception, which seems better than an easy-to-overlook log message.

Reply:
Right, so if it fails to visit file, the exception will be thrown instead of just logging the issue, such as the method is doing. Change will be made to this.